cmd/protoc-gen-grpchan/protoc-gen-grpchan.go

```diff
@@ -5,310 +5,122 @@
-				w("		return nil, err")
-				w("	}")
-				w("	return x, nil")
-				w("}")
+				f.AddElement(gopoet.NewMethod(gopoet.NewPointerReceiverForType("c", cc), mtdName).
+					AddArg("ctx", typeOfContext).
+					AddArg("in", names.GoTypeOfRequest(md)).
+					AddArg("opts", typeOfCallOptions).
+					SetVariadic(true).
+					AddResult("", names.GoTypeForStreamClient(md)).
+					AddResult("", gopoet.ErrorType).
+					Printlnf("stream, err := c.ch.NewStream(ctx, &%s.Streams[%d], \"/%s/%s\", opts...)", names.GoNameOfServiceDesc(sd), streamCount, sd.GetFullyQualifiedName(), md.GetName()).
+					Printlnf("if err != nil {").
+					Printlnf("    return nil, err").
+					Printlnf("}").
+					Printlnf("x := &%s{stream}", names.GoTypeForStreamClientImpl(md)).
+					Printlnf("if err := x.ClientStream.SendMsg(in); err != nil {").
+					Printlnf("    return nil, err").
+					Printlnf("}").
+					Printlnf("if err := x.ClientStream.CloseSend(); err != nil {").
+					Printlnf("    return nil, err").
+					Printlnf("}").
+					Printlnf("return x, nil"))
 				streamCount++
 			} else {
 				// unary method
-				w("func (c *%sChannelClient) %s(ctx context.Context, in *%s%s, opts ...grpc.CallOption) (*%s%s, error) {", lowerSvcName, mtdName, imports.prefixesByFile[md.GetInputType().GetFile()], export(md.GetInputType().GetName()), imports.prefixesByFile[md.GetOutputType().GetFile()], export(md.GetOutputType().GetName()))
-				w("	out := new(%s%s)", imports.prefixesByFile[md.GetOutputType().GetFile()], export(md.GetOutputType().GetName()))
-				w("	err := c.ch.Invoke(ctx, \"/%s/%s\", in, out, opts...)", sd.GetFullyQualifiedName(), md.GetName())
-				w("	if err != nil {")
-				w("		return nil, err")
-				w("	}")
-				w("	return out, nil")
-				w("}")
-			}
-		}
-	}
-
-	file.Content = proto.String(buf.String())
-
-	return &file, nil
-}
-
-func determineFileAndPkg(fd *desc.FileDescriptor) (outputFile, pkg string) {
-	fileName, protoPackage, goOption := fd.GetName(), fd.GetPackage(), fd.GetFileOptions().GetGoPackage()
-	var pkgPath, pkgAlias string
-	if goOption == "" {
-		pkgPath = path.Dir(fileName)
-		if protoPackage == "" {
-			n := path.Base(fileName)
-			ext := path.Ext(n)
-			if ext == "" || len(ext) == len(n) {
-				pkgAlias = n
-			} else {
-				pkgAlias = n[:len(n)-len(ext)]
+				f.AddElement(gopoet.NewMethod(gopoet.NewPointerReceiverForType("c", cc), mtdName).
+					AddArg("ctx", typeOfContext).
+					AddArg("in", names.GoTypeOfRequest(md)).
+					AddArg("opts", typeOfCallOptions).
+					SetVariadic(true).
+					AddResult("", names.GoTypeOfResponse(md)).
+					AddResult("", gopoet.ErrorType).
+					Printlnf("out := new(%s)", names.GoTypeForMessage(md.GetOutputType())).
+					Printlnf("err := c.ch.Invoke(ctx, \"/%s/%s\", in, out, opts...)", sd.GetFullyQualifiedName(), md.GetName()).
+					Printlnf("if err != nil {").
+					Printlnf("    return nil, err").
+					Printlnf("}").
+					Printlnf("return out, nil"))
```

Comment:
This all looks very slick, except for the method body gen.  I wish you could write:

```go
					Printlnf(`
						out := new(%s)
						err := c.ch.Invoke(ctx, "/%s/%s", in, out, opts...)
						if err != nil {
							return nil, err
						}
						return out, nil
					`, names.GoTypeForMessage(md.GetOutputType()), sd.GetFullyQualifiedName(), md.GetName()))

```
and have poet clean up the extra blank lines.

Reply:
Or perhaps even use go's `text/template` in a very neat way.

Reply:
I changed it to use templates (which is something that gopoet supports). WDYT?